### PR TITLE
[BugFix] Fix race condition when access delete predicate list in tablet meta (#42099)

### LINKED ISSUE #42099: [BugFix] Fix race condition when access delete predicate list in tablet meta
tablet::version_for_delete_predicate do not hold lock to access delete predicate list which may cause crash

### PR DESCRIPTION
**Pull request by srlch**
## Why I'm doing:
tablet::version_for_delete_predicate do not hold lock to access delete predicate list which may cause crash.

## What I'm doing:
Try to acquire lock in share mode when access delete predicate list

Fixes #42099

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

File: be/src/storage/tablet.cpp [hunk 1]
```diff
@@ -836,9 +836,19 @@ Status Tablet::_capture_consistent_rowsets_unlocked(const std::vector<Version>&
...
-// version_for_load_deletion()? should at least leave a comment
 bool Tablet::version_for_delete_predicate(const Version& version) {
+    // we use try/catch-block to lock _meta_lock in share mode
+    // It is because in some case, _meta_lock may has been acquire
+    // this the current thread in shared/exclusive mode. It may
+    // be crash if we acquire share lock again. We just do nothing
+    // in this case.
+    try {
+        std::shared_lock rlock(get_header_lock());
+    } catch (const std::exception& e) {}
```
Comment (imay) [inline be/src/storage/tablet.cpp:847]: This lock will be released after this try block, are you sure that this will fix the bug?

Comment (srlch) [reply be/src/storage/tablet.cpp:847]: Yes, it is a problem, maybe we can use get_header_lock().lock_shared() instead. But i still reviewing implementation is reasonable or not. I will fix it later

Comment (srlch) [reply be/src/storage/tablet.cpp]: updated

File: be/src/storage/tablet.cpp [hunk 2]
```diff
@@ -836,9 +838,12 @@ Status Tablet::_capture_consistent_rowsets_unlocked(const std::vector<Version>&
     return Status::OK();
 }
```
Comment (trueeyu) [inline be/src/storage/tablet.cpp:839]: Tablet::get_compaction_status() also need use version_for_delete_predicate_unlocked?

Comment (srlch) [reply be/src/storage/tablet.cpp:839]: updated

**Review by trueeyu — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**